rfc/006-pex-info.md

```diff
@@ -0,0 +1,92 @@
+# RFC 006: Extend PEX messages with additional info
+
+## Changelog
+
+- 2021-07-07: Initial draft
+
+## Author
+
+- @xla
+
+## Context
+
+The majority of tendermint deployments run nodes with varying responsibilities
+offering subsets of all available channels/reactors. While this is accounted
+for in the p2p stack by not sending messages to a peer which did not advise the
+corresponding channel, it depends on the `NodeInfo` exchanged during
+a handshake after a connect. Consequently there currently isn't a protocol
+native way to do rudimentally discovery to find and connect to nodes offering
+certain capabilities.
+
+## Proposal
+
+This document proposes a change to the current set of PEX messages to support
+discovery based on advertised channels. In a way that is backward compabitible
+to avoid it requiring a hard-fork of any running networks.
+
+At the core is the introduction of a new request-response pair which besides
+the endpoint information (ip:port/url) carries the `NodeInfo` exchanged during
+handshake in the response. As that carries the channels active/supported on the
+peer a receiving node can apply filtering appropriate for local use-cases.
+Furthermore any extensions to the `NodeInfo` will be carried with the response
+enabling follow-up changes extending the ABCI to involve the application in the
+node discovery.
+
+### Message changes
+
+The recently introduced [v2 messages](https://github.com/tendermint/spec/pull/312)
+SHALL be replaced with versions which carry additional information about known
+peers.
+
+``` proto
+message PexNode {
+  // Encoded endpoint including protocol, id and path.
+  string   url  = 1 [(gogoproto.customname) = "URL"];
+  // Unaltered NodeInfo obtained during handshake with the peer.
+  NodeInfo info = 2;
+}
+
+message PexRequestInfo {
```

Comment:
Could this message be extended to request peers with certain channels/reactors?

Reply:
Debated this as well out-of-band, @liamsi and I concluded that the complexity is not worth the trade-off until there is a tipping point where the smaller replies are significant in bandwidth saving.

Reply:
As far as I understand the reply contains the node infos, which contain the channels. So you can still do the filtering you want (by channels/reactors), locally on the receiving end.

Reply:
I would say it's mainly a bandwidth saving optimization. 

What is the complexity this would add? Mainly the address book would either order peers accordingly or loop through them?

Reply:
> I would say it's mainly a bandwidth saving optimization.

Yeah that is the angle look at as well if in favour, but I'd like to hear the perspective of node operators on this, as the PEX messages are not frequent afaiu.

> What is the complexity this would add? Mainly the address book would either order peers accordingly or loop through them?

The complexity I'm most vary of his that of the public interface, because beyond a nodes address book now providing and storing the extra info it also has to offer the filtering, which in turn means that needs to be specific and protocolised. Where if it's up to the receiver it can be treated as an implementation detail.